lib/include/openamp/rpmsg.h

```diff
@@ -271,25 +271,20 @@ static inline int rpmsg_trysend_offchannel(struct rpmsg_endpoint *ept,
  * remoteproc address, endpoint callback, and destroy endpoint callback.
  *
  * @ept: pointer to rpmsg endpoint
+ * @rdev: pointer to the rpmsg device
  * @name: service name associated to the endpoint
  * @src: local address of the endpoint
  * @dest: target address of the endpoint
  * @cb: endpoint callback
  * @ns_unbind_cb: end point service unbind callback, called when remote ept is
  *                destroyed.
  */
-static inline void rpmsg_init_ept(struct rpmsg_endpoint *ept,
-				  const char *name,
-				  uint32_t src, uint32_t dest,
-				  rpmsg_ept_cb cb,
-				  rpmsg_ns_unbind_cb ns_unbind_cb)
-{
-	strncpy(ept->name, name ? name : "", sizeof(ept->name));
-	ept->addr = src;
-	ept->dest_addr = dest;
-	ept->cb = cb;
-	ept->ns_unbind_cb = ns_unbind_cb;
-}
+void rpmsg_init_ept(struct rpmsg_endpoint *ept,
+		    struct rpmsg_device *rdev,
+		    const char *name,
+		    uint32_t src, uint32_t dest,
+		    rpmsg_ept_cb cb,
+		    rpmsg_ns_unbind_cb ns_unbind_cb);
```

Comment:
But I can't understand how the user can call the old API without problem: the user can't receive any callback without calling rpmsg_register_endpoint. It isn't only perfect but is wrong! rpmsg_init_ept is designed for the well known server, right? I suppose that all well known server can send/receive message immediately after calling this API.

Reply:
> > But I can't understand how the user can call the old API without problem: the user can't receive any callback without calling rpmsg_register_endpoint.
> 
> Yes it is the normal behavior. you have to register callback for communication.
> 
> > It isn't only perfect but is wrong! rpmsg_init_ept is designed for the well known server, right? I suppose that all well known server can send/receive message immediately after calling this API, but it is impossible without this patch. even worse the user can't call rpmsg_register_endpoint by self because it is declared in a private header file.
> 
> No, rpmsg_init_ept is designed to initialize the ept structure before registering it. But i agree that something is

rpmsg_create_ept initilize all rpmsg_endpoint's fields, why need we call rpmsg_init_ept before that? Does it make sense to initialize the same fields twice?

> not good here.
> The normal API for end point management is rpmsg_create_ept. I can not see a reason why rpmsg_init_ept is exposed. For me this function should be declared in rpmsg_internal.h.
> Any reason that you need to use it, instead of rpmsg_create_ept?

From my interception it's better to:
1.call rpmsg_init_ept for the well known service(both src/dest is fixed) and skip all NS interaction
2.call rpmsg_create_ept for the dynamic serice(either src/dest is any) and utilize NS to find the peer

Reply:
> From my interception it's better to:
> 1.call rpmsg_init_ept for the well known service(both src/dest is fixed) and skip all NS interaction
> 2.call rpmsg_create_ept for the dynamic serice(either src/dest is any) and utilize NS to find the peer

In both case rpmsg_create_ept have to be called, this is what has been implemented since the 2018.10 release. This function handles fixed and non fixed addresses, NS is sent only if destination address is set to RPMSG_ADDR_ANY.
That's why  rpmsg_register_endpoint is not exposed as API  and why rpmsg_init_ept should probably not be exposed.

Reply:
Anyway I just want to make OpenAMP API design become better. If you refuse my change for rpmsg_init_ept I am fine, I will drop in the next update, but for other bug fix

Reply:
I don't think that this point is close. I'm agree with you there is something wrong with rpmsg_init_ept. This is for me a case where we have probably to fix the API.
Firsdt of all i would understand why you propose to duplicate the endpoint API by updating rpmsg_init_ept instead of using rpmsg_create_ept? Any usecase that you can not support using rpmsg_create_ept?

Reply:
I already said before, because rpmsg_create_ept will send NS message, but NS message isn't necessary and may have the bad side effect on kernel side for the well known service because both side already know the peer address. Please imagine how the kernle will response if you create ”NS" endpoint by rpmsg_create_ept.

Reply:
If you think more carefully and spend more time to read OpenAMP code, you can understand the difference between rpmsg_init_ept and rpmsg_create_ept. Anyway, I don't want to spend my time to improve OpenAMP anymore. I just want to fix the bug, so I will always drop this patch in the next update.

Reply:
"NS" channel is created by rpmsg_init_ept, so it's natural to use it for all well known service. Anyway, rpmsg_init_ept in the current form is broken at least as a public API.